### PR TITLE
[TradingMode] add last_activity

### PR DESCRIPTION
**Pull request by GuillaumeDSM**
remaining: tests
requires https://github.com/Drakkar-Software/OctoBot-Trading/pull/1041

### CODE REVIEW

File: Trading/Mode/dca_trading_mode/dca_trading.py [hunk 1]
```diff
@@ -37,6 +37,11 @@
 import octobot_trading.modes.script_keywords as script_keywords
 
 
+class DCAActivity(enum.Enum):
+    CREATED_ORDERS = "created_orders"
+    NOTHING_TO_DO = "nothing_to_do"
```
Comment (Herklos) [inline Trading/Mode/dca_trading_mode/dca_trading.py:42]: Should we use an common enum or it's better to create dedicated keys for DCA ?

Comment (GuillaumeDSM) [reply Trading/Mode/dca_trading_mode/dca_trading.py:42]: we also need to be able to add details on what happend (details on what go rebalanced in indexes for example). I'll change this

File: Trading/Mode/dca_trading_mode/dca_trading.py [hunk 2]
```diff
@@ -440,9 +435,9 @@ async def trigger_dca(self, cryptocurrency, symbol, state):
             f"{symbol} DCA triggered on {self.exchange_manager.exchange_name}, state: {self.state.value}"
         )
         if self.state is trading_enums.EvaluatorStates.NEUTRAL:
-            self.last_activity = DCAActivity.NOTHING_TO_DO
+            self.last_activity = trading_modes.TradingModeActivity(trading_enums.TradingModeActivityType.NOTHING_TO_DO)
```
Comment (Herklos) [inline Trading/Mode/dca_trading_mode/dca_trading.py:438]: :+1:

File: Trading/Mode/index_trading_mode/tests/test_index_trading_mode.py [hunk 3]
```diff
@@ -351,55 +361,206 @@ async def test_should_rebalance(tools):
...
+            get_holdings_ratio_mock.reset_mock()
+        with mock.patch.object(
+            mode, "get_removed_coins_from_previous_config", mock.Mock(return_value=["SOL", "ADA"])
+        ) as get_removed_coins_from_previous_config_mock:
+            should_rebalance, details = producer._get_rebalance_details()
+            assert should_rebalance is True
+            assert details == {
+                index_trading.RebalanceDetails.SELL_SOME.value: {},
+                index_trading.RebalanceDetails.BUY_MORE.value: {},
+                index_trading.RebalanceDetails.REMOVE.value: {
```
Comment (Herklos) [inline Trading/Mode/index_trading_mode/tests/test_index_trading_mode.py:387]: :+1:

File: Trading/Mode/index_trading_mode/index_trading.py [hunk 4]
```diff
@@ -206,25 +261,51 @@ async def _send_alert_notification(self):
...
+            RebalanceDetails.ADD.value: {},
+            RebalanceDetails.SWAP.value: {},
+        }
+        should_rebalance = False
+        # look for coins update in indexed_coins
+        for coin in self.trading_mode.get_removed_coins_from_previous_config():
+            coin_ratio = self.exchange_manager.exchange_personal_data.portfolio_manager. \
+                portfolio_value_holder.get_holdings_ratio(coin, traded_symbols_only=True)
+            if coin_ratio >= trading_constants.ZERO:
+                # coin to sell in portfolio
```
Comment (Herklos) [inline Trading/Mode/index_trading_mode/index_trading.py:278]: :+1:

**Review by Herklos — APPROVED** [review]
Nice work :100:

File: Trading/Mode/index_trading_mode/index_trading.py [hunk 5]
```diff
@@ -198,25 +237,96 @@ async def _send_alert_notification(self):
...
+                rebalance_details[RebalanceDetails.REMOVE.value][origin] * trading_constants.ONE_HUNDRED,
+                3
+            )
+            target_ratio = round(
+                rebalance_details[RebalanceDetails.ADD.value].get(
+                    target,
+                    rebalance_details[RebalanceDetails.BUY_MORE.value].get(target, trading_constants.ZERO)
+                ) * trading_constants.ONE_HUNDRED,
+                3
+            ) or "???"
```
Comment (Herklos) [inline Trading/Mode/index_trading_mode/index_trading.py:298]: Is this value only used for logging purpose ?

Comment (GuillaumeDSM) [reply Trading/Mode/index_trading_mode/index_trading.py:298]: yep

**Review by Herklos — APPROVED** [review]

File: Trading/Mode/index_trading_mode/index_trading.py [hunk 6]
```diff
@@ -32,6 +32,11 @@ class IndexActivity(enum.Enum):
     REBALANCING_SKIPPED = "rebalancing_skipped"
 
 
+class RebalanceSkipDetails(enum.Enum):
+    ALREADY_BALANCED = "already_balanced"
+    NOT_ENOUGH_AVAILABLE_FOUNDS = "not_enough_available_founds"
```
Comment (Herklos) [inline Trading/Mode/index_trading_mode/index_trading.py:37]: :+1:

File: Trading/Mode/index_trading_mode/tests/test_index_trading_mode.py [hunk 7]
```diff
@@ -740,60 +776,105 @@ async def test_split_reference_market_into_indexed_coins(tools):
...
+        # no bought coin
+        details = {index_trading.RebalanceDetails.SWAP.value: {}}
+        mode.indexed_coins = ["ETH", "BTC"]
+        with mock.patch.object(
+            trader.exchange_manager.exchange_personal_data.portfolio_manager.portfolio,
+            "get_currency_portfolio", mock.Mock(return_value=mock.Mock(available=decimal.Decimal("2")))
+        ) as get_currency_portfolio_mock, mock.patch.object(
+            consumer, "_buy_coin", mock.AsyncMock(return_value=[])
+        ) as _buy_coin_mock:
+            with pytest.raises(trading_errors.MissingMinimalExchangeTradeVolume):
```
Comment (Herklos) [inline Trading/Mode/index_trading_mode/tests/test_index_trading_mode.py:827]: :+1:

**Review by Herklos — APPROVED** [review]

**Pull request merged**